htlcswitch/request.go

```diff
@@ -0,0 +1,118 @@
+package htlcswitch
+
+import (
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/routing"
+	"github.com/roasbeef/btcd/wire"
+)
+
+// RequestType represents the type of HTLC switch request type, and it is needed
+// for compilation error in case if wrong type was specified.
+type RequestType uint8
+
+const (
+	// ForwardAddRequest encapsulates the HTLC add request which should be
+	// propagated to another channel over HTLC switch.
+	ForwardAddRequest RequestType = iota
+
+	// UserAddRequest encapsulates the HTLC add request which made by user
+	// and the state of which should be propagated back to user after HTLC
+	// will be settled.
+	UserAddRequest
+
+	// ForwardSettleRequest encapsulates the settle HTLC request which
+	// made by last hope and propagated back to the original hope which sent
+	// the HTLC add request.
+	ForwardSettleRequest
+
+	// CancelRequest encapsulates the cancel HTLC request which propagated
+	// back to the original hope which sent the HTLC add request.
+	CancelRequest
+)
+
+// SwitchRequest encapsulate
+type SwitchRequest struct {
+	// Type is a type of switch request which is used to determine the
+	// necessary behaviour. For example: If HTLC was settled - should we
+	// send the notification over error channel or propagate it back
+	// over HTLC switch?
+	Type RequestType
+
+	// PayHash payment hash of htlc request.
+	PayHash [32]byte
+
+	// Dest is the next peer in HTLC path.
+	Dest *routing.HopID
+
+	// ChannelPoint channel point from which HTLC message came from.
+	ChannelPoint *wire.OutPoint
+
+	// Htlc lnwire HTLC message type of which depends on switch request
+	// type.
+	Htlc lnwire.Message
+
+	// err is an error channel which is used to notify user about
+	// status of payment request is it was canceled or successfully
+	// settled.
+	err chan error
+}
+
+// Error function is used to return request channel error, which make sense
+// only on user request.
+func (r *SwitchRequest) Error() chan error {
+	if r.Type == UserAddRequest {
+		return r.err
+	} else {
+		panic("only user request has error channel")
```

Comment:
`panic` in go is common practice to safe the developer from tricky bugs which may lead to hours and hours of debugging, one of such examples is `time` package. I would rather leave it here in order to prevent developers send error in `nil` channel.